v0.12 testing: A mysterious extra entry

Tests for the `openstack_compute_secgroup_v2` resource are failing. 

You can see the test output here: 

http://logs.openlabtesting.org/logs/59/559/5f55582c3e249ab0d93ef4bf37ffc85e6cfeb384/check/terraform-provider-openstack-acceptance-test/1fdae5a/

via the `job-output.txt.gz` file.

For example, the test `TestAccComputeV2SecGroup_importBasic` has the error message:

```
--- FAIL: TestAccComputeV2SecGroup_importBasic (4.04s)
    testing.go:568: Step 0 error: errors during apply:

        Error: Error creating openstack_compute_secgroup_v2 sg_1 rule: Missing input for argument [IPProtocol]

          on /tmp/tf-test366550116/main.tf line 23:
          (source code not available)
```

The error message is from the Gophercloud library, denoting that the required `IPProtocol` field was omitted. This threw me for a loop because all test fixtures are setting this field (not to mention it passes on v0.11).

After further debugging, I found that there is an empty entry in the [rules set](https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/compute_secgroup_v2.go#L17). Adding a debug line [below this line](https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/compute_secgroup_v2.go#L20) produces:

v0.11.x:

```
2019/02/26 20:37:33 [DEBUG] FOO: map[string]interface {}{"from_port":1, "to_port":65535, "ip_protocol":"udp", "cidr":"0.0.0.0/0", "from_group_id":"", "self":false, "id":""}
2019/02/26 20:37:33 [DEBUG] FOO: map[string]interface {}{"cidr":"0.0.0.0/0", "from_group_id":"", "self":false, "id":"", "from_port":-1, "to_port":-1, "ip_protocol":"icmp"}
2019/02/26 20:37:33 [DEBUG] FOO: map[string]interface {}{"cidr":"0.0.0.0/0", "from_group_id":"", "self":false, "id":"", "from_port":22, "to_port":22, "ip_protocol":"tcp"}
```

v0.12:

```
2019/02/26 20:38:30 [DEBUG] FOO: map[string]interface {}{"to_port":65535, "ip_protocol":"udp", "cidr":"0.0.0.0/0", "from_group_id":"", "self":false, "id":"", "from_port":1}
2019/02/26 20:38:30 [DEBUG] FOO: map[string]interface {}{"ip_protocol":"icmp", "cidr":"0.0.0.0/0", "from_group_id":"", "self":false, "id":"", "from_port":-1, "to_port":-1}
2019/02/26 20:38:30 [DEBUG] FOO: map[string]interface {}{"id":"", "from_port":0, "to_port":0, "ip_protocol":"", "cidr":"0.0.0.0/0", "from_group_id":"", "self":false}
2019/02/26 20:38:30 [DEBUG] FOO: map[string]interface {}{"cidr":"0.0.0.0/0", "from_group_id":"", "self":false, "id":"", "from_port":22, "to_port":22, "ip_protocol":"tcp"}
```

This seems to be the only resource that is seeing this issue. `openstack_compute_secgroup_v2` is one of the oldest and largely unused resources in this provider, so there's probably some bad practices left in over time.

Let me know if I can provide any further details on this one.

--- a/builtin/providers/test/resource_computed_set.go
+++ b/builtin/providers/test/resource_computed_set.go
@@ -1,9 +1,12 @@
 package test
 
 import (
+	"bytes"
 	"fmt"
 	"math/rand"
+	"strings"
 
+	"github.com/hashicorp/terraform/helper/hashcode"
 	"github.com/hashicorp/terraform/helper/schema"
 )
 
@@ -39,8 +42,45 @@ func testResourceComputedSet() *schema.Resource {
 				},
 				Set: schema.HashString,
 			},
+
+			"rule": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+
+						"ip_protocol": {
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: false,
+						},
+
+						"cidr": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: false,
+							StateFunc: func(v interface{}) string {
+								return strings.ToLower(v.(string))
+							},
+						},
+					},
+				},
+			},
 		},
 	}
+}
+
+func computeSecGroupV2RuleHash(v interface{}) int {
+	var buf bytes.Buffer
+	m := v.(map[string]interface{})
+	buf.WriteString(fmt.Sprintf("%s-", m["ip_protocol"].(string)))
+	buf.WriteString(fmt.Sprintf("%s-", strings.ToLower(m["cidr"].(string))))
+
+	return hashcode.String(buf.String())
 }
 
 func testResourceComputedSetCreate(d *schema.ResourceData, meta interface{}) error {

--- a/builtin/providers/test/resource_computed_set_test.go
+++ b/builtin/providers/test/resource_computed_set_test.go
@@ -50,3 +50,22 @@ resource "test_resource_computed_set" "foo" {
 		},
 	})
 }
+
+func TestResourceComputedSet_ruleTest(t *testing.T) {
+	resource.UnitTest(t, resource.TestCase{
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckResourceDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource_computed_set" "foo" {
+	rule {
+		ip_protocol = "udp"
+		cidr = "0.0.0.0/0"
+	}
+}
+				`),
+			},
+		},
+	})
+}

--- a/helper/plugin/grpc_provider.go
+++ b/helper/plugin/grpc_provider.go
@@ -747,6 +747,17 @@ func (s *GRPCProviderServer) ApplyResourceChange(_ context.Context, req *proto.A
 		}
 	}
 
+	// We need to fix any sets that may be using the "~" index prefix to
+	// indicate partially computed. The special sigil isn't really used except
+	// as a clue to visually indicate that the set isn't wholly known.
+	for k, d := range diff.Attributes {
+		if strings.Contains(k, ".~") {
+			delete(diff.Attributes, k)
+			k = strings.Replace(k, ".~", ".", -1)
+			diff.Attributes[k] = d
+		}
+	}
+
 	// add NewExtra Fields that may have been stored in the private data
 	if newExtra := private[newExtraKey]; newExtra != nil {
 		for k, v := range newExtra.(map[string]interface{}) {

